### ISSUE TITLE
 Issues with TLS under write load (again)

### ISSUE THREAD

**Issue by mporsch** [mporsch/sockpuppet#7]
Seems like `ssl3_write_bytes:bad length:ssl` is not leaving us anytime soon, see [test failure in MSVC with OpenSSL 1.1.1u](https://github.com/mporsch/sockpuppet/actions/runs/5769719100/job/15642155542?pr=6).
```
12/14 Test #12: sockpuppet_tls_buffered_test ............***Failed    7.34 sec
generating random reference data
test case #1: unlimited send timeout
server listening at [::1]:8554
client [::1]:50124 connected to server [::1]:8554
sending reference data from [::1]:50124 to [::1]:8554
verifying received against reference data
test case #2: limited send timeout
server listening at [::1]:8554
client [::1]:50125 connected to server [::1]:8554
sending reference data from [::1]:50125 to [::1]:8554
failed to TLS receive/send/handshake: 928:error:1409E10F:SSL routines:ssl3_write_bytes:bad length:ssl\record\rec_layer_s3.c:377:

verifying received against reference data
received only 18876 of 1000000 bytes
test case #3: non-blocking send
server listening at [::1]:8554
client [::1]:50126 connected to server [::1]:8554
sending reference data from [::1]:50126 to [::1]:8554
failed to TLS receive/send/handshake: 2816:error:1409E10F:SSL routines:ssl3_write_bytes:bad length:ssl\record\rec_layer_s3.c:377:

verifying received against reference data
received only 15556 of 1000000 bytes
```
Maybe this is specific to the OpenSSL v1 library used in this github action job's OS? - job has been disabled for now.

**Comment by mporsch** [comment]
Not so specific after all, see [test failure in OSX with OpenSSL 3.1.1](https://github.com/mporsch/sockpuppet/actions/runs/5770396570/job/15643395035)
```
12/14 Test #12: sockpuppet_tls_buffered_test ............***Failed    3.57 sec
generating random reference data
test case #1: unlimited send timeout
server listening at [::1]:8554
client [::1]:49185 connected to server [::1]:8554
sending reference data from [::1]:49185 to [::1]:8554
verifying received against reference data
test case #2: limited send timeout
server listening at [::1]:8554
client [::1]:49186 connected to server [::1]:8554
sending reference data from [::1]:49186 to [::1]:8554
verifying received against reference data
test case #3: non-blocking send
server listening at [::1]:8554
client [::1]:49187 connected to server [::1]:8554
sending reference data from [::1]:49187 to [::1]:8554
failed to TLS receive/send/handshake: 00C0F10B00700000:error:0A00010F:SSL routines:ssl3_write_bytes:bad length:ssl/record/rec_layer_s3.c:370:
```

**Comment by mporsch** [comment]
The `bad length` status was very specific actually: the failing test(s) sample a new TCP packet size after a failed/partial send instead of retrying the previous packet. This breaks the `SSL_write` contract and leads to the error.
Because we already configured `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER` we now check content and length instead of buffer start address.

**Comment by mporsch** [comment]
#5 again?
```
14: Test command: /Users/runner/work/sockpuppet/sockpuppet/build/sockpuppet_tls_async_performance_test
14: Working Directory: /Users/runner/work/sockpuppet/sockpuppet/build
14: Test timeout computed to be: 10000000
14: generating random reference data
14: server listening at 127.0.0.1:49217
14: client 127.0.0.1:49218 connecting to server
14: client 127.0.0.1:49218 connected to server
14: sending reference data from 127.0.0.1:49218 to 127.0.0.1:49217
14: client 127.0.0.1:49219 connecting to server
14: client 127.0.0.1:49219 connected to server
14: sending reference data from 127.0.0.1:49219 to 127.0.0.1:49217
14: client 127.0.0.1:49220 connecting to server
14: client 127.0.0.1:49220 connected to server
14: sending reference data from 127.0.0.1:49220 to 127.0.0.1:49217
14: verifying received against reference data
14: client 127.0.0.1:49218 closing connection to server (self-initiated shutdown)
14: client 127.0.0.1:49218 closed connection to server (failed to TLS receive/send/handshake)
14: verifying received against reference data
14: client 127.0.0.1:49219 closing connection to server (self-initiated shutdown)
14: client 127.0.0.1:49219 closed connection to server (failed to TLS receive/send/handshake)
14: clients did not receive echoed reference data on time
14: client 127.0.0.1:49220 closed connection to server (failed to TLS receive/send/handshake)
14/14 Test #14: sockpuppet_tls_async_performance_test ...***Failed   62.58 sec
```

**Comment by mporsch** [comment]
Fixed by #9
